### PR TITLE
Bug 1254599 app state

### PR DESCRIPTION
**Pull request by fluffyemily**
Using an apparently redux-y design (who knew?), app state.

Currently we have 4 app states, Loading, Tab, Tab Tray and Home Panels, which are described via an enum. The enums can take a state object, which is a struct. The state struct is produced on demand by the component that is describing it's state. This enum is then passed into the menu (or any other components that requires app state to make decisions) so that it can make it's own decisions based on that state. When a components state changes, it notifies via a delegate that the state has changed, passing in its new state as an arguments. The implementer of the delegate then decides how that state should be handled.

Question: Currently I have set up the BrowserViewController to be the app state delegate for the Tab and Home Panels. Should instead, Tab and home panels maintain a list of delegates that it then runs through and notifies, and the Menu is just one of those? I don't really know if I like this idea. If we want to go that way, we are better off sending out an AppStateChanged notification and allowing any components to register as observers. This would enable us to get rid of the delegate entirely. Is this perhaps a better way? I think that what I have implemented here is the more redux-y way of doing things, but maybe I am wrong.

Other changes in this PR:

    Total refactor or Browser, renaming it, and anything else that is actually tab related rather than related to the entire app, Tab
    Turning MenuConfiguration, MenuItem and MenuToolbarItem into protocols in readyness for potential future separation of Menu from main project (as part of the whole can we Open Source any of our components question)

N.B.: a t the current time, the tests I wrote for Menu are failing. I'm not going to update those until I know we are happy with the approach for AppState and then I shall also add some AppState tests.

### CODE REVIEW

**Comment by sleroux** [comment]
This is looking really good overall! I'm excited to see the 'state' lifted out of our classes and into structs. I think this will make it easier to reason about our app.

> Question: Currently I have set up the BrowserViewController to be the app state delegate for the Tab and Home Panels. Should instead, Tab and home panels maintain a list of delegates that it then runs through and notifies, and the Menu is just one of those? I don't really know if I like this idea. If we want to go that way, we are better off sending out an AppStateChanged notification and allowing any components to register as observers

I think for future-proofing I can see using a NSNotification to be the better option. This way we don't need to worry about linking disparate components of the app together with a delegate. I also tend to avoid using delegates for 1->N relationships since it can get unruly. As long as we're careful about when and how we register/unregister the observers I think this will give us the best flexibility.

One minor nit is just the naming of the concrete implementation of the MenuItem to FirefoxMenuItem. I really like the idea of lifting the protocol out to allow for mocking, etc but I think FirefoxMenuItem is also too generic of a name. Maybe something more specific such as PopoverMenuItem, ChromeMenuItem? Not sure what makes sense there.

**Comment by fluffyemily** [comment]
naming things was never my strong point. Happy to rename anything to
something more sensible.

File: Client/Frontend/Browser/Tab.swift [hunk 1]
```diff
@@ -417,13 +451,24 @@ class Browser: NSObject, BrowserWebViewDelegate {
         }
     }
 
-    private func browserWebView(browserWebView: BrowserWebView, didSelectFindInPageForSelection selection: String) {
-        browserDelegate?.browser(self, didSelectFindInPageForSelection: selection)
+    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String: AnyObject]?, context: UnsafeMutablePointer<Void>) {
+        guard let webView = object as? WKWebView where webView == self.webView,
+            let path = keyPath where path == "URL"else {
```
Comment (thebnich) [inline Client/Frontend/Browser/Tab.swift]: Nit: space before `else`

File: Client/Frontend/Menu/MenuConfiguration.swift [hunk 2]
```diff
@@ -4,146 +4,171 @@
...
-    case HomePanels
-    case TabTray
+protocol MenuConfiguration {
+
+    var menuItems: [MenuItem] { get }
+    var menuToolbarItems: [MenuToolbarItem]? { get }
+    var numberOfItemsInRow: Int { get }
+
+    init(appState: AppState)
+    func toolbarColour() -> UIColor
```
Comment (thebnich) [inline Client/Frontend/Menu/MenuConfiguration.swift]: Nit: `Colour` -> `Color` for consistency

**Comment by thebnich** [comment]
Sorry for the late review -- suddenly had a lot come up at once, and I'm still catching up! Just dumping my thoughts here:

On Android, each activity has an [`onCreateOptionsMenu`](https://developer.android.com/reference/android/app/Activity.html#onCreateOptionsMenu(android.view.Menu)) method that takes a Menu object and populates it with items specific to that activity. Translating that to iOS, this design would have `BrowserViewController`, `TabTrayController`, and `HomePanelViewController` each implement a `MenuProvider` protocol such that each of these VCs directly updates the menu state to match the VC state.

With your approach, VCs instead build objects containing the minimum necessary state needed for the menu to make its decisions, and `MenuConfiguration` does all of the heavy lifting translating this state into UI. There's more coupling with this implementation since each VC that wants to participate in menu handling needs to provide a (somewhat arbitrary) VC-specific state object. The Android approach might be a bit more open-source-friendly since (IIUC) apps wanting to use the menu code in this PR with multiple VCs will need to reimplement a similar state system. On the other hand, it's nice that the menu code is all in one place, and you're able to reuse shared actions such as New Tab and Settings.

So, comparisons aside, LGTM! It'll be interesting to see how this pans out over time compared to the menu we're using on Android.

File: Client/Frontend/Menu/MenuConfiguration.swift [hunk 3]
```diff
@@ -4,146 +4,171 @@
...
+    func toolbarTintColor() -> UIColor
+    func menuBackgroundColor() -> UIColor
+    func menuTintColor() -> UIColor
+    func menuFont() -> UIFont
+    func menuIcon() -> UIImage?
+    func shadowColor() -> UIColor
 }
 
-struct MenuConfiguration {
+struct FirefoxMenuConfiguration: MenuConfiguration {
```
Comment (thebnich) [inline Client/Frontend/Menu/MenuConfiguration.swift]: Historically, we've avoided using product names (Firefox) directly in code. We already have the `App` prefix (`AppConstants` and now `AppState`), so perhaps `AppMenuItem`/`AppMenuConfiguration` would work here, too?

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -1135,25 +1138,44 @@ class BrowserViewController: UIViewController {
...
+    }
 }
 
 /**
  * History visit management.
  * TODO: this should be expanded to track various visit types; see Bug 1166084.
  */
 extension BrowserViewController {
-    func ignoreNavigationInTab(tab: Browser, navigation: WKNavigation) {
+    func ignoreNavigationInTab(tab: Tab, navigation: WKNavigation) {
```
Comment (jhugman) [inline Client/Frontend/Browser/BrowserViewController.swift]: At what point do we make `BrowserViewController` a container for a `TabViewController` and a tool bar?

**Pull request merged**